### PR TITLE
Add Typed Events for CalDAV- and CardDAV-related Events

### PR DESCRIPTION
**Pull request by georgehrke**
* [x] document new events in https://docs.nextcloud.com/server/19/developer_manual/app/events.html#available-events
* [x] add to #20953 a note about the deprecation

| Deprecated | Use instead |
|:---------:|:------:|
|`\OCA\DAV\CalDAV\CalDavBackend::createCalendar`|`OCA\DAV\Events\CalendarCreatedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::updateCalendar`|`OCA\DAV\Events\CalendarUpdatedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::deleteCalendar`|`OCA\DAV\Events\CalendarDeletedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::createCalendarObject`|`OCA\DAV\Events\CalendarObjectCreatedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::updateCalendarObject`|`OCA\DAV\Events\CalendarObjectUpdatedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::deleteCalendarObject`|`OCA\DAV\Events\CalendarObjectDeletedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::createSubscription`|`OCA\DAV\Events\SubscriptionCreatedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::updateSubscription`|`OCA\DAV\Events\SubscriptionUpdatedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::deleteSubscription`|`OCA\DAV\Events\SubscriptionDeletedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::createCachedCalendarObject`|`OCA\DAV\Events\CachedCalendarObjectCreatedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::updateCachedCalendarObject`|`OCA\DAV\Events\CachedCalendarObjectUpdatedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::deleteCachedCalendarObject`|`OCA\DAV\Events\CachedCalendarObjectDeletedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::updateShares`|`OCA\DAV\Events\CalendarShareUpdatedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::publishCalendar`|`OCA\DAV\Events\CalendarPublishedEvent`|
|`\OCA\DAV\CalDAV\CalDavBackend::publishCalendar`|`OCA\DAV\Events\CalendarUnpublishedEvent`|
|`\OCA\DAV\CardDAV\CardDavBackend::createCard`|`OCA\DAV\Events\CardCreatedEvent`|
|`\OCA\DAV\CardDAV\CardDavBackend::updateCard`|`OCA\DAV\Events\CardUpdatedEvent`|
|`\OCA\DAV\CardDAV\CardDavBackend::deleteCard`|`OCA\DAV\Events\CardDeletedEvent`|

Documentation PR: https://github.com/nextcloud/documentation/pull/2765

### CODE REVIEW

**Comment by georgehrke** [comment]
How long should we keep the old events in place?

**Comment by skjnldsv** [comment]
> How long should we keep the old events in place?

usual deprecation? 3 releases?

**Comment by georgehrke** [comment]
CalDAV tester is not happy, need to look into that

**Comment by georgehrke** [comment]
> CalDAV tester is not happy, need to look into that

fixed

**Comment by MorrisJobke** [comment]
> How long should we keep the old events in place?

Put them also into #20953 and see #14552.

**Review by MorrisJobke — APPROVED** [review]
Fine with me 👍

**Comment by MorrisJobke** [comment]
Also the documentation should be updated - I added it to the OP.

**Comment by MorrisJobke** [comment]
@georgehrke Mind to look into the failing events?

**Comment by MorrisJobke** [comment]
Looks far better now 👍

**Review by rullzer — APPROVED** [review]
:rocket:

**Review by skjnldsv — APPROVED** [review]

**Comment by MorrisJobke** [comment]
@georgehrke I assume it's ready to merge, right?

File: apps/dav/lib/CardDAV/CardDavBackend.php [hunk 1]
```diff
@@ -388,6 +402,10 @@ public function updateAddressBook($addressBookId, \Sabre\DAV\PropPatch $propPatc
 
 			$this->addChange($addressBookId, "", 2);
 
+			$addressBookRow = $this->getAddressBookById($addressBookId);
```
Comment (MorrisJobke) [inline apps/dav/lib/CardDAV/CardDavBackend.php:405]: `InvalidScalarArgument: Argument 1 of OCA\DAV\CardDAV\CardDavBackend::getAddressBookById expects int, string provided`

Comment (georgehrke) [reply apps/dav/lib/CardDAV/CardDavBackend.php:405]: Looking into that now

Comment (MorrisJobke) [reply apps/dav/lib/CardDAV/CardDavBackend.php]: Works now 👍

**Review by MorrisJobke — APPROVED** [review]
Together with #22304 psalm is also happy 👍

**Comment by MorrisJobke** [comment]
Merged the documentation and added it to the critical changes ticket.

**Pull request merged**